src/Features/CSharp/Portable/Completion/KeywordRecommenders/StackAllocKeywordRecommender.cs

```diff
@@ -16,20 +16,48 @@ public StackAllocKeywordRecommender()
 
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
-            // type t = |
-            var token = context.TargetToken;
-            if (token.IsUnsafeContext())
+            var node = context.TargetToken.Parent;
+
+            // In case of an empty string
+            if (node == null)
+            {
+                return false;
+            }
+
+            // After a cast
+            if (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) &&
+                (node.IsKind(SyntaxKind.ParenthesizedExpression) || node.IsKind(SyntaxKind.CastExpression)))
+            {
+                node = node.Parent;
+            }
+
+            // Inside a conditional expression: value ? stackalloc : stackalloc
+            while (node.IsKind(SyntaxKind.ConditionalExpression) &&
+                (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) || context.TargetToken.IsKind(SyntaxKind.QuestionToken) || context.TargetToken.IsKind(SyntaxKind.ColonToken)))
+            {
+                node = node.Parent;
+            }
+
+            // assignment: x = stackalloc
+            if (node.IsKind(SyntaxKind.SimpleAssignmentExpression))
```

Comment:
should check that you're after the = just to be paranoid.

Also, please add test that you don't get stackalloc here:

```x $$ = ```

Reply:
We might not be after equal. We might be nested inside a conditional or a cast :)
The root node has to be an assignment, but not necessarily the direct parent.

---
In reply to: [156192877](https://github.com/dotnet/roslyn/pull/23655#discussion_r156192877) [](ancestors = 156192877)

Reply:
This should then validate that you're on the right side of the equal.

Basically, your'e validating yourself against a certain shape of the tree and the position of the user.  We need to make sure that that shape is tight and doesn't allow for other cases like this.